### ISSUE TITLE
port-forward should be able to write its port bindings to a file then background itself

### ISSUE THREAD

**Issue by ringerc** [kubernetes/kubectl#1190]
**What would you like to be added**:

A command-line option for `kubectl port-forward` to write a list bound ports and its pid to, then daemonize itself only once successfully bound.

Currently scripting `kubectl port-forward` is a real pain, especially if you want to handle errors and/or use dynamic local ports with `:portname` format.

**Why is this needed**:

This would allow scripts to use `kubectl port-forward` *much* more easily.

Currently scripts need to interact with `kubectl port-forward` like a co-process, reading its stdout/stderr until it prints the list of bound ports, then parsing them from its output, checking it's still running, and handling any errors if it exits. This is clumsy and error-prone.

With an option to write ports to a file or file-like object scripts could easily test kubectl's exit code, and if successful, know that the port-forward is ready to use.

Here's how it might look by hand:

```
$ kubectl port-forward --daemonize --ports-file=service-foo-portname.json --ports-file-format=json service/foo :portname
Forwarding from 127.0.0.1:42777 -> 9411
Forwarding from [::1]:42777 -> 9411
Wrote ports to service-foo-portname.json
Port forwarding established, daemonized with pid 295942
$ echo $?
0
$ cat service-foo-portname.json
{
  "pid": "295942",
  "forwards": [
    {
      "protocol": "ipv4",
      "remote": {
        "port": "9411"
      },
      "local": {
        "addr": "0.0.0.0",
        "port": "42777",
        "addr_and_port": "0.0.0.0:42777"
      }
    },
    {
      "protocol": "ipv6",
      "remote": {
        "port": "9411"
      },
      "local": {
        "addr": "::1",
        "port": "42777",
        "addr_and_port": "[::1]:42777"
      }
    }
  ]
}
```

Here's how it might look in `bash` with `--ports-file` writing to `stdout` using the convention of `-` as the filename:

```sh
if kubectl_ports=$(kubectl port-forward --daemonize --ports-file=- --ports-file-format=json service/foo :portname)
then
  local_addr=$(jq -r '.|first|.local.addr_and_port' <<<$kubectl_ports)
  kubectl_pid=$(jq -r '.pid' <<<$kubectl_ports)
  echo "bound to ${local_addr}"
else
  errcode=$?
  echo "ERROR: kubectl port-forward exited with code $errcode, could not establish port forwards"
  exit $errcode
fi

run_my_program ${local_addr}

kill $kubectl_pid
```

(It might be possible to retain `kubectl` under shell job control instead of daemonizing, which would be even nicer since the shell would retain process tree ownership and still kill it on exit.)

**Comment by ringerc** [comment]
By contrast here's some bash code I wrote to use `kubectl port-forward` as it stands now.

```
# Start a kubectl port-forward and wait until the port is active
#
# The local port and local bind address is automatically selected and is
# returned as global variable
#
#    port_forward_address
#
function start_port_forward() {
    coproc $kubectl port-forward service/upm-api-dp-provisioner :grpc -n upm-api-dp-provisioner </dev/null 2>&1
    port_forward_pid=$COPROC_PID
    while IFS='' read -r -t $PORT_FORWARD_TIMEOUT_SECONDS -u "${COPROC[0]}" LINE
    do
        if [[ "$LINE" == *"Forwarding from"* ]]; then
            port_forward_address="${LINE#Forwarding from *}"   
            port_forward_address="${port_forward_address%% -> *}"
            if [ -z "${port_forward_address}" ]; then
              echo "ERROR: Failed to get local address for port-forward"
              echo "kubectl output line: $LINE"
              exit 1
            fi
            # Remaining output is on stderr, which we don't capture, so we
            # should be fine to ignore the stdout file descriptor now and
            # port_forward_pid remains set and will be used on cleanup
            #
            return
        else    
            echo "kubectl: ${LINE}"
        fi
    done
    # if we reached here, read failed, likely due to the coproc exiting
    if [ -n "${port_forward_pid:-}" ]; then
      port_forward_ecode=
      wait $port_forward_pid || port_forward_ecode=$?
      echo "port-forward request failed? Exit code $port_forward_ecode"
    else
      echo "port forward request failed? Could not get kubectl port-forward's pid"
    fi
    exit 1
}
```

In addition to being really ugly, it's fragile as it relies on parsing kubectl's human readable output.

**Comment by ringerc** [comment]
It'd be reasonable to separate this into two issues, too:

* Write the ports to a json file
* Daemonize after port-forwarding is established

**Comment by eddiezane** [comment]
I don't think daemonizing is something we'd want to add but a plugin would be a good place to start with a proof of concept.

/triage needs-plugin-prototype

**Comment by ringerc** [comment]
I went to bang out a minimal `kubectl` patch for this, since it seemed like a bit of a no-brainer to support writing a json document listing the ports to a file / stdout.

It looks like https://github.com/kubernetes/client-go/blob/master/tools/portforward/portforward.go has a `GetPorts()` method that will block until ready, so I added a wrapper for that to the mockable wrapper class. Then waited until the port mapping was ready and wrote the resulting file.

It's not beautiful and needs cleaning up, but it should work: https://github.com/ringerc/kubectl/tree/write-port-mapping-file commit https://github.com/ringerc/kubectl/commit/31ac025f2f11fb1416fc495ebcd9502b1a07c41c .

I'm not sure whether I'll go through the process to try to upstream this against kubectl staging yet. I suspect the patch will need refinement if I do. But anyway, here's a patch demonstrating how it can be done, along with tests showing it works.
